convert2rhel/actions/__init__.py

```diff
@@ -0,0 +1,619 @@
+    #                 "message": "" or "$message"},
+    # }
+    formatted_results = {}
+    for action in itertools.chain(*results):
+        formatted_results[action.id] = {"status": action.status, "error_id": action.error_id, "message": action.message}
+    return formatted_results
+
+
+def find_actions_of_severity(results, severity):
+    """
+    Filter results from :func:`run_actions` to include only results of ``severity`` or higher.
+
+    :param results: Results dictionary as returned by :func:`run_actions`
+    :type results: Mapping
+    :param severity: The name of a ``STATUS_CODE`` for the severity to filter to.
+    :returns: List of actions which are at ``severity`` or higher result. Empty list
+        if there were no failures.
+    :rtype: Sequence
+
+    Example::
+
+        matched_actions = find_actions_of_severity(results, "SKIP")
+        # matched_actions will contain all actions which were skipped
+        # or failed while running.
+    """
+    matched_actions = [action for action in results.items() if action[1]["status"] >= STATUS_CODE[severity]]
+
+    return matched_actions
+
+
+def format_report_section_heading(status_code):
+    """
+    Format a section heading for a status in the report.
+
+    :param status_code: The status code that will be used in the heading
+    :type status_code: int
+    :return: The formatted heading that the caller can log.
+    :rtype: str
+    """
+    status_header = _STATUS_HEADER[status_code]
+    highlight = "=" * 10
+
+    heading = "{highlight} {status_header} {highlight}".format(highlight=highlight, status_header=status_header)
+    return heading
+
+
+def format_report_message(status_code, action_id, error_id, message):
+    """Helper function to format a message about each Action result.
+
+    :param status_code: The status code that will be used in the template.
+    :type status_code: int
+    :param action_id: Action id for the message
+    :type action_id: str
+    :param error_id: Error id associated with the action
+    :type error_id: str
+    :param message: The message that was produced in the action
+    :type message: str
+
+    :return: The formatted message that will be logged to the user.
+    :rtype: str
+    """
+    status_name = _STATUS_NAME_FROM_CODE[status_code]
+    template = "({STATUS}) {ACTION_ID}"
+    # `error_id` and `message` may not be present everytime, since it
+    # can be empty (either by mistake, or, intended), we only want to
+    # apply these fields if they are present, with a special mention to
+    # `message`.
+    if error_id:
+        template += ".{ERROR_ID}"
+
+    # Special case for `message` to not output empty message to the
+    # user without message.
+    if message:
+        template += ": {MESSAGE}"
+    else:
+        template += ": [No further information given]"
+
+    return template.format(
+        STATUS=status_name,
+        ACTION_ID=action_id,
+        ERROR_ID=error_id,
+        MESSAGE=message,
+    )
```

Comment:
Since those two functions are only being used inside the report module, shouldn't we move then there? It seems to me that they would belong there instead of here. 

Maybe moving `_STATUS_NAME_FROM_CODE`  too, not sure...

Reply:
These two functions are here because of circular imports.  But your mention of `STATUS_NAME_FROM_CODE` might be able to untangle that.  I'll take a look at see if there's a way to move all the relevant code to report and post here about the problem if I can't find a way to do it.

Reply:
* `format_report_message()` is used by `Stage.run()`.
  * It's used there because we decided that Actions which error should print out a status message right after they run in addition to inside of the report.
  * I propose changing the name of this to `format_action_status_message()` so it's more clear that it's tied to action instead of report.
  * `_STATUS_NAME_FROM_CODE` needs to remain here if this function remains here to avoid circular import problems.
* I don't see a circular import reason for `format_report_section_header()` and `_STATUS_HEADER` to be in `action/__init__.py`.
  * My guess is I put `_STATUS_HEADER` there because it uses status code codes (integers) which are explained by what's in `STATUS_CODE`.  I think I will leave it here for that reason.
  * I probably put `format_report_section_header()` here because I put `format_report_message()` here.  But renaming `format_report_message()` to `format_action_status_message()` makes it clear I was mistaken to think those functions belonged together.  The section headers are only useful for the report so it belongs in report.py while the formatted status message is useful anywhere that Actions might be useful.

Reply:
Done.  Let me know if you think there's more that can be done here.

Reply:
This was implemented in this commit: f965c4aa591914c1742fa97aaee306f4f442f925